shade-test/jar-file/src/test/java/org/testcontainers/PublicAPITest.java

```diff
@@ -0,0 +1,120 @@
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@RunWith(Parameterized.class)
+@RequiredArgsConstructor
+public class PublicAPITest extends AbstractJarFileTest {
+
+    private static String SHADED_PACKAGE = "org.testcontainers.shaded.";
+    private static String SHADED_PACKAGE_PATH = SHADED_PACKAGE.replaceAll("\\.", "/");
+
+    static {
+        Assertions.registerFormatterForType(ClassNode.class, it -> it.name);
+        Assertions.registerFormatterForType(FieldNode.class, it -> it.name);
+        Assertions.registerFormatterForType(MethodNode.class, it -> it.name + it.desc);
+    }
+
+    @Parameters(name = "{0}")
+    public static List<Object[]> data() throws Exception {
+        List<Object[]> result = new ArrayList<>();
+
+        Files.walkFileTree(root, new SimpleFileVisitor<Path>() {
+
+            @Override
+            public FileVisitResult visitFile(Path path, BasicFileAttributes attrs) throws IOException {
+                String fileName = path.toString();
+
+                if (!fileName.endsWith(".class")) {
+                    return super.visitFile(path, attrs);
+                }
+
+                if (!fileName.startsWith("/org/testcontainers/") ) {
+                     return super.visitFile(path, attrs);
+                }
+
+                if (fileName.startsWith("/" + SHADED_PACKAGE_PATH)) {
+                    return super.visitFile(path, attrs);
+                }
+
+                try(InputStream inputStream = Files.newInputStream(path)) {
+                    ClassReader reader = new ClassReader(inputStream);
+                    ClassNode node = new ClassNode();
+                    reader.accept(node, ClassReader.SKIP_CODE);
+                    if ((node.access & Opcodes.ACC_PUBLIC) != 0) {
+                        result.add(new Object[]{fileName, node});
+                    }
+                }
+
+                return super.visitFile(path, attrs);
+            }
+        });
+        return result;
+    }
+
+    private final String fileName;
+
+    private final ClassNode classNode;
+
+    @Test
+    public void testSuperClass() {
+        assertThat(classNode.superName)
+                .doesNotStartWith(SHADED_PACKAGE_PATH);
+    }
+
+    @Test
+    public void testInterfaces() {
+        assertThat(classNode.interfaces)
+                .allSatisfy(it -> assertThat(it).doesNotStartWith(SHADED_PACKAGE_PATH));
+    }
+
+    @Test
+    public void testMethodReturnTypes() {
+        assertThat(classNode.methods)
+                .filteredOn(it -> (it.access & (Opcodes.ACC_PUBLIC | Opcodes.ACC_PROTECTED)) != 0)
+                .allSatisfy(it -> assertThat(Type.getReturnType(it.desc).getClassName()).doesNotStartWith(SHADED_PACKAGE));
+    }
+
+    @Test
+    public void testMethodArguments() {
+        assertThat(classNode.methods)
```

Comment:
I was about to comment on how I could not understand the simplest thing: the `withCreateContainerCmdModifier` method on `GenericContainer`, and how it passes this test. This is the one place where we consciously leak the docker-java API - so when it's shaded, the class that it this method accepts must surely include the relocated package name. 

I was struggling to understand this when I realized: type erasure 🙀. The signature is this:

    public SELF withCreateContainerCmdModifier(Consumer<CreateContainerCmd> modifier)

which resolves to a type description of:

    (Ljava/util/function/Consumer;)Lorg/testcontainers/containers/GenericContainer;

This doesn't fall foul of the checks here, because generics.

For the record, I think this one slipping through is fine because this is a conscious API leak, and you probably realised this already 😄 

For other methods that involve generic arguments, can you think of any risk areas that we should carefully look at manually (since reflection won't be able to see through the generic type)?